Fixing visualization code for object detection to support both types of bounding box.

Anindyadeep opened the pull request:
# What does this PR do?

In the current documentation of using [transformers for object detection](https://huggingface.co/docs/transformers/tasks/object_detection). There belongs a section where we visualize the bounding boxes. But when the same code is used in a different dataset with un-normalized bounding box, it does not work. For a new comer, or someone trying to see results with different dataset might cause friction. So, this PR tries to cover up that edge case.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #27817


## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Documentation: @stevhliu and @MKhalusova

Anindyadeep committed:
Merge pull request #1 from huggingface/main

merge from main.

Commit:
fix: minor enhancement and fix in bounding box visualization example

The example that was trying to visualize the bounding box was not considering an edge case,
where the bounding box can be un-normalized. So using the same set of code, we can not get
results with a different dataset with un-normalized bounding box. This commit fixes that.

MKhalusova commented:
Thank you for your contribution! Please make sure the CI checks are green: run `make style` to fix the issues.

Anindyadeep commented:
> Thank you for your contribution! Please make sure the CI checks are green: run `make style` to fix the issues.

Yeah, so I did that, and doing provides changes in other files. Here:

```
modified:   src/transformers/models/bloom/modeling_bloom.py
modified:   src/transformers/models/fuyu/image_processing_fuyu.py
modified:   src/transformers/models/mpt/modeling_mpt.py
```

So, should I still add it?

Anindyadeep committed:
Merge pull request #2 from huggingface/main

merge from main.

Commit:
run make clean

Commit:
Merge branch 'anindya/fix-od-visualization' of https://github.com/Anindyadeep/transformers into anindya/fix-od-visualization

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_27842). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

MKhalusova commented:
Thank you for making the example more robust to different datasets! LGTM

ArthurZucker commented on src/transformers/models/bloom/modeling_bloom.py:
these are unrelated changed make sure you have `ruff==0.1.5` and run make style~ 🤗

Anindyadeep replied:
> these are unrelated changed make sure you have `ruff==0.1.5` and run make style~ 🤗

Yep, I ran `make style` with `ruff==0.1.5`

ArthurZucker reviewed (commented):
LGTM but let's remove unrelated changes!

Anindyadeep committed:
Merge pull request #3 from huggingface/main

merge from main.

Commit:
Merge branch 'main' into anindya/fix-od-visualization

amyeroberts commented on docs/source/en/tasks/object_detection.md:
```diff
@@ -139,6 +139,17 @@ To get an even better understanding of the data, visualize an example in the dat
 ...     box = annotations["bbox"][i]
 ...     class_idx = annotations["category"][i]
 ...     x, y, w, h = tuple(box)
+...     # Check if coordinates are normalized or not
+...     if max(box) > 1.0:
+...         # Coordinates are un-normalized, no need to re-scale them
+...         x1, y1 = int(x), int(y)
+...         x2, y2 = int(x + w), int(y + h)
+...     else:
+...         # Coordinates are normalized, re-scale them
+...         x1 = int(x * width)
+...         y1 = int(y * height)
+...         x2 = int((x + w) * width)
```
The difficulty with this is, even with this change, the bounding box visualization still won't be fully general as there's different bounding box formats e.g. (x, y, w, h) vs (x1, y1, x2, y2). 

Instead of trying to handle this all in a single code example, what I would suggest is adding a note to mention the format that this code works for and that one might have to convert their bounding boxes to match.

Anindyadeep replied:
Got it, so should I remove this, and add a documentation change with a note, or should I keep this code, and add a say something like:

"This above piece of code tries to handle for certain formats, and here we are trying to handle cases for un-normalized or normalized, however, this is not generalized for all kinds of format ... " And then I mention some example of other formats?

amyeroberts replied:
Let's keep the code here as it's useful for users to know how to map from relative to absolute. You can just have a note in the documentation above saying that it assumes XYWH format, which might be different for some models and datasets.

Anindyadeep replied:
Sounds great

amyeroberts reviewed (commented):
Thanks for adding this!

Small comment about handling this on the code side

Anindyadeep committed:
Merge pull request #4 from huggingface/main

Merge from main.

Commit:
Merge branch 'main' into anindya/fix-od-visualization

Commit:
add an additional note on the scenarios where the box viz code works

amyeroberts reviewed (approved):
Awesome - thanks for adding this and iterating!
